Lib/zipfile.py

```diff
@@ -1818,7 +1818,9 @@ def close(self):
                              "Close the writing handle before closing the zip.")
 
         try:
-            if self.mode in ('w', 'x', 'a') and self._didModify: # write ending records
+            fp_open = not self.fp.closed if isinstance(self.fp, io.IOBase) else True
+
+            if self.mode in ('w', 'x', 'a') and self._didModify and fp_open: # write ending records
```

Comment:
Is there a risk here of corrupting the user's intention? If the user has modified the zip file but the end record hasn't been written, couldn't that leave the zip file incomplete? I would expect if one closes the zip file before it's been finalized that it _should_ fail.